lib/services/proxywatcher.go

```diff
@@ -0,0 +1,278 @@
+		return trace.Wrap(err)
+	}
+	defer watcher.Close()
+	reloadC := time.After(p.ReloadPeriod)
+	// before fetch, make sure watcher is synced by receiving init event,
+	// to avoid the scenario:
+	// 1. Cache process:   w = NewWatcher()
+	// 2. Cache process:   c.fetch()
+	// 3. Backend process: addItem()
+	// 4. Cache process:   <- w.Events()
+	//
+	// If there is a way that NewWatcher() on line 1 could
+	// return without subscription established first,
+	// Code line 3 could execute and line 4 could miss event,
+	// wrapping up with out of sync replica.
+	// To avoid this, before doing fetch,
+	// cache process makes sure the connection is established
+	// by receiving init event first.
+	select {
+	case <-watcher.Done():
+		return trace.ConnectionProblem(watcher.Error(), "watcher is closed")
+	case <-reloadC:
+		p.Debugf("Triggering scheduled reload.")
+		return nil
+	case <-p.Context.Done():
+		return trace.ConnectionProblem(p.Context.Err(), "context is closing")
+	case event := <-watcher.Events():
+		if event.Type != backend.OpInit {
+			return trace.BadParameter("expected init event, got %v instead", event.Type)
+		}
+	}
+	proxies, err := p.Client.GetProxies()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	proxySet := make(map[string]Server, len(proxies))
+	for i := range proxies {
+		proxySet[proxies[i].GetName()] = proxies[i]
+	}
+	p.retry.Reset()
+	p.setCurrent(proxies)
+	select {
+	case p.ProxiesC <- proxies:
+	case <-p.Context.Done():
+		return trace.ConnectionProblem(p.Context.Err(), "context is closing")
+	}
+	for {
+		select {
+		case <-watcher.Done():
+			return trace.ConnectionProblem(watcher.Error(), "watcher is closed")
+		case <-reloadC:
+			p.Debugf("Triggering scheduled reload.")
+			return nil
+		case <-p.Context.Done():
+			return trace.ConnectionProblem(p.Context.Err(), "context is closing")
+		case event := <-watcher.Events():
+			updated := p.processEvent(event, proxySet)
+			if updated {
+				proxies = setToList(proxySet)
+				p.setCurrent(proxies)
+				select {
+				case p.ProxiesC <- proxies:
+				case <-p.Context.Done():
+					return trace.ConnectionProblem(p.Context.Err(), "context is closing")
+				}
+			}
+		}
+	}
+}
+
+// processEvent updates proxy map and returns true if the proxies list have been modified -
+// the proxy has been either added or deleted
+func (p *ProxyWatcher) processEvent(event Event, proxies map[string]Server) bool {
+	if event.Resource.GetKind() != KindProxy {
+		p.Warningf("Unexpected event: %v.")
+		return false
+	}
+	switch event.Type {
+	case backend.OpDelete:
+		delete(proxies, event.Resource.GetName())
+		// always return true if the proxy has been deleted to trigger
+		// broadcast cleanup
+		return true
+	case backend.OpPut:
+		_, existed := proxies[event.Resource.GetName()]
+		resource, ok := event.Resource.(Server)
+		if !ok {
+			p.Warningf("unexpected type %T", event.Resource)
+			return false
+		}
+		proxies[event.Resource.GetName()] = resource
+		return !existed
```

Comment:
If the existing proxy is updated, this will return "false" and thus won't trigger reload AFAIU. Only if the proxy didn't exist before. Is that intended behavior, i.e. shouldn't reload also happen if existing proxy is updated?

Reply:
yes, this is intended